### PR TITLE
nexus: improve optimal_tilematrix routine

### PR DESCRIPTION
**Pull request by jtkrogel**
optimal_tilematrix (structure.py) is improved in the following ways:
1) Accelerated computation of candidate tilematrix determinants for target volume matching.
2) Ability to use a mask array to restrict form of tilematrix search, e.g. only look for tilings of the form:
  x x 0
  x x 0
  0 0 x
3) Use of wigner radius to select among cells w/ equivalent inscribing radius.
4) Use of devation of cubic shape (as judged by face diagonals) to select among cells w/ equivalent wigner radius.
5) Prioritization of certain tile matrix aesthetics (diagonal, symmetric, etc) to select among cells w/ equivalent cubicity

Changes result in speedup of ~5x and superior cell selection for typical problem sets over the original algorithm.

Other minor updates include a fix for when the user provides no PhysicalSystem to a generate simulation function, the addition of a function to read an arbitrary input file and return the appropriate SimulationInput object, improved handling namespaces for qmcpack xml elements, and the ability to generate energy density xml elements with a more succinct UI.

### CODE REVIEW

File: nexus/library/qmcpack_input.py [hunk 1]
```diff
@@ -5049,6 +5056,122 @@ def count_jastrow_params(jastrows):
...
+        name      = None,
+        dynamic   = None,
+        static    = None,
+        coord     = None,
+        grid      = None,
+        scale     = None,
+        ion_grids = None,
+        system    = None,
+        ):
+    type = 'EnergyDensity'
```
Comment (markdewing) [inline nexus/library/qmcpack_input.py]: This value is unused (from pyflakes analysis)

**Comment by jtkrogel** [comment]
Thanks.  The spurious line has been removed.

**Comment by markdewing** [comment]
Could you fix the __main__ code in structure.py?  The arguments to generate_structure must have changed since it was written.

What is optimal_tilematrix supposed to do?  (Take the input axes and fit them into a cubic simulation box? Or fit <something> into a given volume? )

**Comment by jtkrogel** [comment]
I've cleaned up what was in __main__ for structure.py.  The code there looked like it was circa 2012 before I decided to broadly use keyword arguments.  The "fold" function called there is deprecated; it is replaced entirely by tiling methods.

optimal_tilematrix is meant to find the best cell for QMC purposes that is a fixed factor (say 4x) larger in volume than the current (likely primitive) cell.  It tries to find one with the largest wigner radius (to maximize the range of our short-ranged jastrows) that is also most cubically shaped (to minimize finite size effects and/or improve the quality of finite size corrections).

**Comment by prckent** [comment]
@jtkrogel Some tests for nexus would be desirable to help prove that the code works/still works. e.g. a few workflows of unconverged calculations to generated reasonable coverage.

**Comment by jtkrogel** [comment]
Agreed.  I will make a github issue for further discussion on this.

**Pull request merged**